tfhe/src/keycache/mod.rs

```diff
@@ -186,32 +196,34 @@ pub mod utils {
                         .map(|param_key| param_key.1.clone())
                 };
 
-                if let Some(shared_cell) = maybe_shared_cell {
-                    shared_cell.inner.get_or_init(load_from_persistent_storage);
-                    Ok(shared_cell)
-                } else {
-                    Err(())
-                }
+                maybe_shared_cell.map_or_else(
+                    || {
+                        {
+                            // we only hold a write lock for a short duration to push the lazily
+                            // evaluated key without actually evaluating the key
```

Comment:
that comment does not look accurate, the code below does the generation while holding the lock.

I think we should check if a key is in there, if it's in there then done, otherwise generate it and then re-acquire the lock to write in there

In practice it makes little difference for us for now as we pre-generate keys, but I tjink I would prefer have this done once and for all